iox_query/src/exec/gapfill.rs

```diff
@@ -604,4 +614,244 @@ mod test {
+            for i in 0..self.group_cols.len() {
+                fields.push(Field::new(
+                    &format!("g{i}"),
+                    DataType::Dictionary(Box::new(DataType::Int32), Box::new(DataType::Utf8)),
+                    true,
+                ));
+            }
+            fields.push(Field::new(
+                "t",
+                DataType::Timestamp(TimeUnit::Nanosecond, None),
+                true,
+            ));
+            for i in 0..self.agg_cols.len() {
+                fields.push(Field::new(&format!("a{i}"), DataType::Int32, true));
+            }
+            Arc::new(Schema::new(fields))
+        }
+
+        fn exprs(&self) -> Result<(ExprVec, ExprVec)> {
+            let mut group_expr: ExprVec = vec![];
+            let mut aggr_expr: ExprVec = vec![];
+            let ngroup_cols = self.group_cols.len();
+            for i in 0..self.schema().fields().len() {
+                match i.cmp(&ngroup_cols) {
+                    Ordering::Less => group_expr.push(Arc::new(Column::new(&format!("g{i}"), i))),
+                    Ordering::Equal => group_expr.push(Arc::new(Column::new("t", i))),
+                    Ordering::Greater => {
+                        let idx = i - ngroup_cols + 1;
+                        aggr_expr.push(Arc::new(Column::new(&format!("a{idx}"), i)));
+                    }
+                }
+            }
+            Ok((group_expr, aggr_expr))
+        }
+    }
+
+    impl TryFrom<TestBatch> for RecordBatch {
+        type Error = DataFusionError;
+
+        fn try_from(value: TestBatch) -> Result<Self> {
+            let mut arrs: Vec<ArrayRef> =
+                Vec::with_capacity(value.group_cols.len() + value.agg_cols.len() + 1);
+            for gc in &value.group_cols {
+                let arr = Arc::new(DictionaryArray::<Int32Type>::from_iter(gc.iter().cloned()));
+                arrs.push(arr);
+            }
+            // Scale from milliseconds to the nanoseconds that are actually stored.
+            let scaled_times: Vec<Option<i64>> = value
+                .time_col
+                .iter()
+                .map(|o| o.map(|v| v * 1_000_000))
+                .collect();
+            arrs.push(Arc::new(TimestampNanosecondArray::from(scaled_times)));
+            for ac in &value.agg_cols {
+                let arr = Arc::new(Int32Array::from_iter(ac));
+                arrs.push(arr);
+            }
+
+            Self::try_new(value.schema(), arrs).map_err(DataFusionError::ArrowError)
+        }
+    }
+
+    struct TestCase {
+        in_batches: Vec<TestBatch>,
+        batch_size: usize,
+        params: GapFillExecParams,
+    }
+
+    impl TestCase {
+        async fn run(mut self) -> Result<Vec<RecordBatch>> {
+            let schema = self.in_batches[0].schema();
+            let (group_expr, aggr_expr) = self.in_batches[0].exprs()?;
+            let mut test_batches = vec![];
+            std::mem::swap(&mut test_batches, &mut self.in_batches);
```

Comment:
I wonder why you had to do this swap? I think you should just be able to use self.in_batches directly


Like

```suggestion
            let mut test_batches = self.in_batches;
```